### ISSUE TITLE
Random Feedback

### ISSUE THREAD

**Issue by Kipt** [josefreichelt/klasik-games#1]
Howdy, as you asked for feedback on Game Dev League discord, here are my 10 cents.

**TL;DR**  No big issues as far as I can tell, all looks good in general - but some minor details and nitpickings.

1. Very minor potential issue, Apple frameworks.

CMake and Apple frameworks, according to [CMake Wiki](https://gitlab.kitware.com/cmake/community/-/wikis/doc/cmake/platform_dependent_issues/HowToUseExistingOSXFrameworks) suggests
```
find_library(framework_IOKit IOKit)
target_link_libraries(${PROJECT_NAME} PRIVATE ${framework_IOKit})
```
instead of just `-framework IOKit` as is used [here](https://github.com/josefreichelt/klasik-games/blob/main/pong/CMakeLists.txt#L41). I'm not sure if there's much or any difference in practice, hence a very minor issue if at all.

2. Potential minor annoyance in future, lacking namespace.

Your types and functions are in global namespace. This _can_ lead to issues in the future as including C++ header files will also "leak in" the C header files that are not part of `std::` namespace. Say for example you add a function called `log`. As C also defines this function you might end up with multiple definitions issue. One fix that is very common practice is to have everything (except `main`) inside a namespace, such as `namespace klasik { ... }`.

3. Potential minor annoyance, `int main`.

Although the entry point is correct according to C standard, for games and non-console applications on Windows: `WinMain` (or `wWinMain`) is preferred. If you see a console popping up when you start the game it's because the subsystem is console and not GUI. See [msdn](https://learn.microsoft.com/en-us/windows/win32/learnwin32/winmain--the-application-entry-point) for reference.

4. Issue with globals, at `main.cpp`.

Currently not an issue, but the global constants defined [here](https://github.com/josefreichelt/klasik-games/blob/main/pong/source/main.cpp#L8) should be defined as `static` or inside an anonymous namespace (preferred by many C++ developers). The constants are currently defined as global, so if you in another translation unit (i.e. another .cpp file) define a constant with the same name, you'll get a linking error (violating the One-Definition-Rule, ODR). Specifying `static` or putting them in an anonymous namespace tells the compiler to not expose the constant outside the translation unit and thus can only be used inside that .cpp file.

5. Issue with usage of window dimensions.

The playable area seems to be in pixels depending on the window dimensions. This might cause issues. Say you are in the middle of the game and resize the window, suddenly the game area changes (and I suspect the ball would appear to move left).

6. Minor issue, `Ball::size` ambiguity.

Better name might be `radius` as `size` might be interpreted as diameter.

7. Very minor note, debug output to console.

Of course it's not a finished product you're shipping to Steam players, but `printf` and `std::cout` for debug output is usually not the best way. I would add a debug output function and use `DebugOutputA` on Windows to get debug messages directly to debugger output.

8. Dogmatic C++ notes.

Although it matters very little, some C++ developers get really dogmatic about using C++ versions of C functionality. Some of usages here are:
- C-style casts could be changed to C++-style [here](https://github.com/josefreichelt/klasik-games/blob/main/pong/source/ball.cpp#L51).
- Usage of C `rand` [here](https://github.com/josefreichelt/klasik-games/blob/main/pong/source/ball.cpp#L45) instead of C++ randomization.
- Usage of `printf` [here](https://github.com/josefreichelt/klasik-games/blob/main/pong/source/main.cpp#L53).

9. Minor detail, suppression of Clang / GCC enum-switch warning.

Both Clang and GCC have warnings for when you switch on an enum value and you don't have a case for it. This is suppressed when adding a `default` case as is done [here](https://github.com/josefreichelt/klasik-games/blob/main/pong/source/main.cpp#L138). I suggest either removing the `default` case, or putting a runtime assertion that the default branch is never hit.

10. Include guards.

As all compilers you'll most likely compile C++ on nowadays supports `#pragma once`, you could use that to save some typing. But of course, if you want to be 100% standard compliant include guards are the way to go.

11. Minor detail, missing `srand`.

Currently usage of `rand` will produce the exact same values each time you run the game, probably not intended. Add `srand` call to seed the random generator.

12. Minor detail, sub-optimal randomization.

The code [here](https://github.com/josefreichelt/klasik-games/blob/main/pong/source/ball.cpp#L43) could be simplified to
```
if (rand() % 2 == 0)
{
    new_direction_x = -1;
}
else
{
    new_direction_x = 1;
}
```
This prevents a loop that might randomly take a long time.

**Comment by josefreichelt** [comment]
Thank you so much for this feedback, will try to incorporate these improvements in future games.
